internal_ws/yktrace/src/swt.rs

```diff
@@ -2,74 +2,49 @@
 
 use super::{SirTrace, ThreadTracer, ThreadTracerImpl};
 use crate::{errors::InvalidTraceError, sir::SirLoc};
-use libc;
-use std::convert::TryFrom;
+use std::cell::RefCell;
 
 /// Softare thread tracer.
 struct SWTThreadTracer;
 
 /// Stop tracing on the current thread.
 impl ThreadTracerImpl for SWTThreadTracer {
     fn stop_tracing(&mut self) -> Result<SirTrace, InvalidTraceError> {
-        let len;
-        let buf = unsafe {
-            if !__YK_SWT_ACTIVE {
-                libc::puts(
-                    "attempted to stop tracing when no tracer is active\n\0" as *const str
-                        as *const i8
-                );
-                libc::free(TRACE_BUF as *mut _);
-                TRACE_BUF = 0 as *mut SwtLoc;
-                TRACE_BUF_LEN = 0;
-                return Err(InvalidTraceError::InternalError);
-            }
+        if unsafe { !__YK_SWT_ACTIVE } {
+            println!("attempted to stop tracing when no tracer is active");
+            return Err(InvalidTraceError::InternalError);
+        }
 
+        unsafe {
             __YK_SWT_ACTIVE = false;
+        }
+
+        let locs = TRACE_BUF.with(|trace_buf| {
+            // When we make a SirTrace, we convert all of the locations from SwtLoc to SirLoc.
+            trace_buf
+                .borrow_mut()
+                .drain(..)
+                .map(|swt_loc| {
+                    let symbol_name = unsafe { std::ffi::CStr::from_ptr(swt_loc.symbol_name) };
+                    SirLoc {
+                        symbol_name: symbol_name.to_str().unwrap(),
+                        bb_idx: swt_loc.bb_idx,
+                        addr: None
+                    }
+                })
+                .collect()
+        });
 
-            // We hand ownership of the trace to the caller. The caller is responsible
-            // for freeing the trace.
-            let ret_trace = TRACE_BUF;
-            len = TRACE_BUF_LEN;
-
-            // Now reset all off the recorder's state.
-            TRACE_BUF = 0 as *mut SwtLoc;
```

Comment:
I think I got this right. The above `drain()` means this stuff isn't needed?

Reply:
Correct